crates/lib/src/action/worker.rs

```diff
@@ -32,17 +32,132 @@ pub async fn worker(
 	events_tx: priority::Sender<Event, Priority>,
 	events: priority::Receiver<Event, Priority>,
 ) -> Result<(), CriticalError> {
-	let mut last = Instant::now();
-	let mut set = Vec::new();
-	let process = ProcessHolder::default();
-	let outcome_gen = OutcomeWorker::newgen();
+	let mut processes: HashMap<SupervisorId, ProcessData> = HashMap::new();
+
+	// Insert the base process from the initial `WorkingData`.
+	let base_process_id = SupervisorId::default();
+	processes.insert(
+		base_process_id,
+		ProcessData {
+			working: working.clone(),
+			process: ProcessHolder::default(),
+			outcome_gen: OutcomeWorker::newgen(),
+		},
+	);
+
+	while let Some(mut set) = throttle_collect(
+		events.clone(),
+		working.clone(),
+		errors.clone(),
+		Instant::now(),
+	)
+	.await?
+	{
+		#[allow(clippy::iter_with_drain)]
+		let events = Arc::from(set.drain(..).collect::<Vec<_>>().into_boxed_slice());
+		let action = Action::new(
+			Arc::clone(&events),
+			Arc::from(
+				processes
+					.keys()
+					.copied()
+					.collect::<Vec<_>>()
+					.into_boxed_slice(),
+			),
+		);
+		debug!("running action handler");
+		let action_handler = {
+			let wrk = working.borrow();
+			wrk.action_handler.clone()
+		};
 
-	loop {
-		if events.is_closed() {
-			trace!("events channel closed, stopping");
-			break;
+		let outcomes = action.outcomes.clone();
+		let err = action_handler
+			.call(action)
+			.await
+			.map_err(|e| rte("action worker", e.as_ref()));
+		if let Err(err) = err {
+			errors.send(err).await?;
+			debug!("action handler errored, skipping");
+			continue;
 		}
 
+		for (pid, (resolution, event_set)) in outcomes.lock().await.iter() {
+			let found = match resolution {
+				Resolution::Apply(outcome) => processes
+					.get(pid)
+					.map(|data| (outcome.clone(), data.clone())),
+				Resolution::Start(cmd) => {
+					assert!(processes.get(pid).is_none());
+					let mut wrk = working.borrow().clone();
```

Comment:
oh, i see! 

would you prefer the lock being acquired as a variable and then having that variable dropped explicitly to ensure this is what is happening?

otherwise, by comment, would you mean the mention of this as in:
```rust
// due to borrow semantics, lock is only held for this line
```

Reply:
That comment would be good yes.